File: ws/first_ws/src/ctrl_turtle_pkg/src/ctrl_turtle.cpp
//
// Created by zcb on 2021/4/3.
//

#include <iostream>
// ros
#include <ros/ros.h>
#include <geometry_msgs/Twist.h>

// qt
#include <QApplication>
#include <QWidget>
#include <QFormLayout>
#include <QLineEdit>
#include <QPushButton>

using namespace std;

ros::Publisher publisher;
QLineEdit *editLinear, *editAngular;

void slotFuc() {
    geometry_msgs::Twist msg;

    double linearX = editLinear->text().toDouble();
    double angularZ = editAngular->text().toDouble();

    msg.linear.x = linearX;
    msg.angular.z = angularZ * M_PI / 180;

    publisher.publish(msg);
}

int main(int argc, char **argv) {

    //1 node
    string nodeName = "ctrl_turtle_cpp_node";
    ros::init(argc, argv, nodeName);
    ros::NodeHandle node;

    string topicName = "/turtle1/cmd_vel";
    publisher = node.advertise<geometry_msgs::Twist>(topicName, 1000);

    //2 qt
    QApplication app(argc, argv);
    QWidget window;
    window.setWindowTitle("小乌龟控制器");

    // layout
    QFormLayout formLayout(&window);

    editLinear = new QLineEdit("0.0");
    editAngular = new QLineEdit("0.0");
    formLayout.addRow("线速度", editLinear);
    formLayout.addRow("角速度", editAngular);

    QPushButton sendBtn("发送");
    formLayout.addRow(&sendBtn);

    // event:sendBtn
    QPushButton::connect(&sendBtn, &QPushButton::clicked, &slotFuc);

    window.show();

    return app.exec();
}

File: 06ros/ws01/first_ws/src/ctrl_turtle_pkg/scripts/ctrl_turtle.py
#! /usr/bin/env python
# coding:utf8
import sys
import rospy
from PyQt5.QtWidgets import (
    QWidget, QApplication, QFormLayout, QLineEdit, QPushButton,
)
from geometry_msgs.msg import Twist


def send_msg():
    global publisher, linearXEdit, angularZEdit
    msg = Twist()
    linearX = float(linearXEdit.text())
    angularZ = float(angularZEdit.text())
    msg.linear.x = linearX
    msg.angular.z = angularZ * 3.14 / 180
    publisher.publish(msg)


if __name__ == '__main__':
    # ros
    nodeName = "ctrl_turtle_py_node"
    rospy.init_node(nodeName)
    publisher = rospy.Publisher("/turtle1/cmd_vel", Twist, queue_size=1000)

    # qt
    app = QApplication(sys.argv)
    window = QWidget()
    window.setWindowTitle("小乌龟速度控制")
    layout = QFormLayout(window)

    linearXEdit = QLineEdit("0.0")
    angularZEdit = QLineEdit("0.0")
    layout.addRow("线速度", linearXEdit)
    layout.addRow("角速度", angularZEdit)

    sendBtn = QPushButton("发送")
    layout.addRow(sendBtn)

    # slot func
    sendBtn.clicked.connect(send_msg)

    window.show()

    sys.exit(app.exec_())
